balancer-js/src/modules/pools/pool-types/concerns/metaStable/priceImpact.concern.ts

```diff
@@ -0,0 +1,67 @@
+import { PriceImpactConcern } from '@/modules/pools/pool-types/concerns/types';
+import { calcPriceImpact } from '@/modules/pricing/priceImpact';
+import {
+  ONE,
+  BZERO,
+  SolidityMaths,
+  _upscale,
+  _computeScalingFactor,
+} from '@/lib/utils/solidityMaths';
+import { BalancerError, BalancerErrorCode } from '@/balancerErrors';
+import { Pool } from '@/types';
+import { parsePoolInfo } from '@/lib/utils';
+import { bptSpotPrice } from '@/lib/utils/stableMathHelpers';
+
+export class MetaStablePoolPriceImpact implements PriceImpactConcern {
+  /**
+   * Calculates the BPT return amount when investing with no price impact.
+   * @param { Pool } pool Investment pool.
+   * @param { string [] } amounts Token amounts being invested. Needs a value for each pool token.
+   * @returns { string } BPT amount.
+   */
+  bptZeroPriceImpact(pool: Pool, tokenAmounts: bigint[]): bigint {
+    if (tokenAmounts.length !== pool.tokensList.length)
+      throw new BalancerError(BalancerErrorCode.INPUT_LENGTH_MISMATCH);
+
+    // const tokensList = cloneDeep(pool.tokensList);
+    const { parsedBalances, parsedPriceRates, parsedAmp, parsedTotalShares } =
```

Comment:
I know it shouldn't happen, but it would be nice to have a sanity check and throw errors if `parsedAmp` and `parsedPriceRates` are undefined. Then you'll be able to use them without casting their types.